tfjs-backend-wasm/src/kernels/Dilation2D.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {backend_util, Dilation2D, Dilation2DAttrs, Dilation2DInputs, KernelConfig, KernelFunc, TensorInfo} from '@tensorflow/tfjs-core';
+
+import {BackendWasm} from '../backend_wasm';
+import {CppDType} from './types';
+
+let wasmDilation2D: (
+    xId: number, filterId: number, outId: number, dtype: number, batch: number,
+    depth: number, inHeight: number, inWidth: number, outHeight: number,
+    outWidth: number, strideHeight: number, strideWidth: number,
+    dilationHeight: number, dilationWidth: number, filterHeight: number,
+    filterWidth: number, padTop: number, padLeft: number) => void;
+
+function setup(backend: BackendWasm) {
+  wasmDilation2D = backend.wasm.cwrap(Dilation2D, null, [
+    'number',  // xId
+    'number',  // filterId
+    'number',  // outId
+    'number',  // dtype
+    'number',  // batch
+    'number',  // depth
+    'number',  // inHeight
+    'number',  // inWidth
+    'number',  // outHeight
+    'number',  // outWidth
+    'number',  // strideHeight
+    'number',  // strideWidth
+    'number',  // dilationHeight
+    'number',  // dilationWidth
+    'number',  // filterHeight
+    'number',  // filterWidth
+    'number',  // padTop
+    'number',  // padLeft
+  ]);
+}
+
+export function dilation2D(args: {
+  inputs: Dilation2DInputs,
+  attrs: Dilation2DAttrs,
+  backend: BackendWasm,
+}): TensorInfo {
+  const {inputs, backend, attrs} = args;
+  const {x, filter} = inputs;
+  const {strides, pad, dilations} = attrs;
+
+  if (x.dtype !== filter.dtype) {
```

Comment:
Could we move this check to core? As I know, we typically do users' input check in core; otherwise the backends' behaviors could be different.

Reply:
This is intentional: our cpu and webgl backend can take different types and do the type cast automatically, while wasm backend is very hard to do that. Therefore this check is just here to make the wasm implementation easier.

The TF Dilation2D op also has this constraint that x and filter must have the same type, so it should be fine in terms of correctness. However, it will be a breaking change to TFJS if we move this check to core level.

Reply:
Thanks for explanation!